src/kernel/notifications_interface.h

```diff
@@ -27,6 +28,7 @@ class Notifications
     virtual void headerTip(SynchronizationState state, int64_t height, int64_t timestamp, bool presync) {}
     virtual void progress(const std::string& title, int progress_percent, bool resume_possible) {}
     virtual void warning(const bilingual_str& warning) {}
+    virtual void fatalError(const std::string& debug_message, bilingual_str user_message = {}) {}
```

Comment:
In commit "kernel: Add fatalError method to notifications" (0b88c307a8ed81705cf8e6fb6332fdf969eb0e2e)

Would be good to change `bilingual_str` to `const bilingual_str&`. That should also avoid the need to remove the forward declaration and include `#include <util/translation.h>` above.

Reply:
Not sure about this one. The include is required for setting the default argument and it is non-const, because we mutate the message.

Reply:
> Not sure about this one. The include is required for setting the default argument and it is non-const, because we mutate the message.

That's a good point about the include. But changing to a `const bilingual_str&` reference and avoiding unnecessary copies would provide a more consistent interface.

For one thing, getting of the mutations would actually make current implementations of the interface more legible:

```c++
InitError(user_message.empty() ? _("A fatal internal error occurred, see debug.log for details") : user_message);

std::cerr << (user_message.empty() ? "A fatal internal error occurred." : user_message.original) << std::endl;
```

But more importantly kernel code should not care about quirks of particular implementations of the interface. Only 2 out of the 3 of the fatalError implementation currently mutate their arguments, and mutating the argument is always unnecessary. I don't think you would argue it would make sense to change other `const string&` and `const bilingual_str&` parameters in the interface if some future implementation of those interfaces decides it wants to be quirky and mutate arguments. The interface should pass strings by reference or by value in some consistent way.

Reply:
Thanks for the follow-up!